### ISSUE TITLE
[HOLD for payment 2024-06-06] [$250]  IOU - "Members" appears in IOU details page and then disappears after a few seconds

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#41832]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.71-0
**Reproducible in staging?:** Yes
**Reproducible in production?:** No
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause Internal Team
**Slack conversation:**

## Action Performed:
1. Login as user A to staging.new.expensify.com
2. Submit two expenses to B
3. Press IOU preview to open transaction page
4. Select any IOU request
5. Press header to open the details page
## Expected Result:
Only settings is seen in details page of IOU request.
## Actual Result:
"Members" is seen briefly and then disappears

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [X] iOS: mWeb Safari
- [X] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/38435837/412f06e8-2c0b-4d80-8fcb-0f393b12ea33

https://github.com/Expensify/App/assets/38435837/962bb69c-2bca-4326-b5d8-b19adb5316b6

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~012a67c3a384fec2d0</li>
        <li>Upwork Job ID: 1788211386365521920</li>
        <li>Last Price Increase: 2024-05-08</li>
<li>Automatic offers: </li>
<ul>
<li>Pujan92 | Reviewer | 0</li>
<li>nkdengineer | Contributor | 0</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @trjExpensify</details>

**Comment by m-natarajan** [comment]
@rlinoz FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by rlinoz** [comment]
oh, that was considered a bug, and we enabled the tap now.

**Comment by rlinoz** [comment]
Ok, so I think this is a problem that we had before, it just wasn't possible to click on.

>I'll be curious to follow the git blame to see where we made the changes to not show Members in iouReport/iouTransaction details.

That is hard to follow, we changed how participants work a lot recently, but I don't know if we should have ever shown members here, since the notification preference for transactions is `hidden`. Now we are actively filtering out hidden members, so that is why it is not showing the members section.

The problem of showing then hiding is happening because we are optimistically setting the members as not hidden for this types of reports. We get the participants [here](https://github.com/Expensify/App/blob/e0b6a06d2f34f35cda96ccc4d05252ede111c732/src/libs/ReportUtils.ts#L4783) and always set them as `hidden:false` [here](https://github.com/Expensify/App/blob/e0b6a06d2f34f35cda96ccc4d05252ede111c732/src/libs/ReportUtils.ts#L4296).

@trjExpensify but I am curious, do you think we should bypass the notification preference here and show the participants?

**Comment by trjExpensify** [comment]
>The problem of showing then hiding is happening because we are optimistically setting the members as not hidden for this types of reports. We get the participants [here](https://github.com/Expensify/App/blob/e0b6a06d2f34f35cda96ccc4d05252ede111c732/src/libs/ReportUtils.ts#L4783) and always set them as hidden:false [here](https://github.com/Expensify/App/blob/e0b6a06d2f34f35cda96ccc4d05252ede111c732/src/libs/ReportUtils.ts#L4296).

Ahh, gotcha. So until the report is unhidden, members don't show. Gotcha! That makes sense, I guess. 👍

**Comment by rlinoz** [comment]
Nice, I will remove the deploy blocker label as this problem and the members section being hidden is the correct behavior, and open it up for proposals.

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
"Members" is seen briefly and then disappears

### What is the root cause of that problem?
When generating optimistic chat report for the transaction thread, we're setting `hidden: false` for the participants [here](https://github.com/Expensify/App/blob/e0b6a06d2f34f35cda96ccc4d05252ede111c732/src/libs/ReportUtils.ts#L4296), similar to other types of chat reports.

This is not right because the `participants` for transaction thread should be hidden, when the back-end response comes, it will have `hidden: true` correctly. That's why the `Members` only shows briefly, because the optimistic data is wrong and the back-end data returned is correct.

### What changes do you think we should make in order to solve the problem?

When generating optimistic chat report for the transaction thread, we should always have `hidden: true` for the participants (including the case of `OwnPolicyExpenseChat` too because in that case the `Members` also should not show). To do this we can:
- Add another param `shouldShowParticipants` to `buildOptimisticChatReport`, the default value is `true`
- Update [this](https://github.com/Expensify/App/blob/e0b6a06d2f34f35cda96ccc4d05252ede111c732/src/libs/ReportUtils.ts#L4296) to
```
hidden: !shouldShowParticipants,
```
So if `shouldShowParticipants` is `true`, `hidden` will be false and vice versa.
- When generating optimistic chat report for transaction thread [here](https://github.com/Expensify/App/blob/e0b6a06d2f34f35cda96ccc4d05252ede111c732/src/libs/ReportUtils.ts#L4797), pass `shouldShowParticipants` as `false`

If there's other report types where `buildOptimisticChatReport` should not show participants, also pass `shouldShowParticipants` as `false`

### What alternative solutions did you explore? (Optional)
Instead of using a new param `shouldShowParticipants`, we can try to detect that the current report is a transaction thread inside `buildOptimisticChatReport` and set `hidden` accordingly. But having a param `shouldShowParticipants` seems clearer and also works for other cases where `shouldShowParticipants` should be `true`

**Comment by Pujan92** [comment]
@rlinoz To confirm, for all transaction threads hidden will be true from the backend? Bcoz if that is the case then we should update the optimistic value specifically in the front end.

**Comment by rlinoz** [comment]
Yes, for every new transaction thread every member will be hidden until they perform some action on it.

**Comment by Pujan92** [comment]
Thanks @rlinoz.

@GandalfGwaihir and @cretadn22's main solution proposed to update the condition inside the `ReportDetailsPage` component, but I think it's better to update optimistic data for the transaction threads correctly which will be consistent with backend. 

I prefer to move with @nkdengineer's [proposal](https://github.com/Expensify/App/issues/41832#issuecomment-2101024857) where for any of the transaction threads we should set `hidden` to true for participants.

🎀👀🎀 C+ reviewed

**Comment by nkdengineer** [comment]
@Pujan92 The PR is [here](https://github.com/Expensify/App/pull/42036).

**Comment by Pujan92** [comment]
**BugZero Checklist**: The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [X] [@Pujan92] The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/App/pull/37458
- [X] [@Pujan92] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: https://github.com/Expensify/App/pull/37458/files#r1629616409
 - [X] [@Pujan92] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: NA
 - [X] [@Pujan92] Determine if we should create a regression test for this bug. No
 - [X] [@Pujan92] If we decide to create a [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) for the bug, please propose the regression test steps to ensure the same bug will not reach production again.

**Comment by trjExpensify** [comment]
Payment summary as follows: 

- $250 to @nkdengineer for the fix
- $250 to @Pujan92 for the C+ review 

Paid!
